### PR TITLE
Fix: hidden items in item label based focus

### PR DESCRIPTION
**Pull request by hongduc-phan**
Hidden Items can be focused with tabIndex = 0.

I fixed it by checking the hidden condition, if the item is hidden, it should be not focused and tabIndex is -1.
The number of issue: #64

### CODE REVIEW

File: vaadin-list-mixin.html [hunk 1]
```diff
@@ -208,18 +210,27 @@
...
-      idx = this._getAvailableIndex(idx, 1, item => !item.disabled);
+      idx = this._getAvailableIndex(idx, 1, item => !(item.disabled || getComputedStyle(item).getPropertyValue('display') === 'none'));
       const item = this.items[idx] || this.items[0];
       this.items.forEach(e => e.tabIndex = e === item ? 0 : -1);
     }
 
     _focus(idx) {
       const item = this.items[idx];
-      this.items.forEach(e => e.focused = e === item);
+      this._setHidden(item);
```
Comment (tomivirkki) [inline vaadin-list-mixin.html]: This shouldn't be here. `hidden` shouldn't be modified by the component. Looks like `hidden` is also never removed if the `display: none` is removed from an item

File: vaadin-list-mixin.html [hunk 2]
```diff
@@ -208,18 +210,27 @@
...
 
+    _setHidden(item) {
+      if (getComputedStyle(item).getPropertyValue('display') === 'none') {
+        item.hidden = true;
+      }
+    }
+
     _setFocusable(idx) {
-      idx = this._getAvailableIndex(idx, 1, item => !item.disabled);
+      idx = this._getAvailableIndex(idx, 1, item => !(item.disabled || getComputedStyle(item).getPropertyValue('display') === 'none'));
```
Comment (tomivirkki) [inline vaadin-list-mixin.html]: Instead of repeating `getComputedStyle(item).getPropertyValue('display') === 'none'` in multiple places, it's probably better to create a `_isHidden(item)` method that checks it instead

File: vaadin-list-mixin.html [hunk 3]
```diff
@@ -208,18 +210,27 @@
...
 
     _focus(idx) {
       const item = this.items[idx];
-      this.items.forEach(e => e.focused = e === item);
+      this._setHidden(item);
+      this.items.forEach(e => e.focused = (e === item && !(e.hidden || e.disabled)));
       this._setFocusable(idx);
-      this._scrollToItem(idx);
-      item.focus();
+      if (!(item.hidden || item.disabled)) {
```
Comment (tomivirkki) [inline vaadin-list-mixin.html]: Rather than fixing the logic in `_focus` it might be better to do it where `_focus` gets invoked (in `_onKeydown`, like the previous fix)

File: vaadin-list-mixin.html (hunk 2 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done

File: vaadin-list-mixin.html (hunk 1 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done

File: vaadin-list-mixin.html (hunk 3 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done

File: vaadin-list-mixin.html [hunk 4]
```diff
@@ -208,15 +208,22 @@
       return -1;
     }
 
+    _isItemHidden (item) {
+      if (getComputedStyle(item).getPropertyValue('display') === 'none') {
+        return true;
+      }
+      return false;
```
Comment (tomivirkki) [inline vaadin-list-mixin.html]: `return getComputedStyle(item).getPropertyValue('display') === 'none';`

File: vaadin-list-mixin.html [hunk 5]
```diff
@@ -208,15 +208,22 @@
...
-      idx = this._getAvailableIndex(idx, 1, item => !item.disabled);
+      idx = this._getAvailableIndex(idx, 1, item => !(item.disabled || this._isItemHidden(item)));
       const item = this.items[idx] || this.items[0];
       this.items.forEach(e => e.tabIndex = e === item ? 0 : -1);
     }
 
     _focus(idx) {
       const item = this.items[idx];
-      this.items.forEach(e => e.focused = e === item);
+      this.items.forEach(e => e.focused = (e === item && !(this._isItemHidden(e) || e.disabled)));
```
Comment (tomivirkki) [inline vaadin-list-mixin.html]: This change might not be needed (if the condition is already checked elsewhere so you can never reach this block with a hidden item index). If there are no other side effects, it might actually make more sense to have all the validity checks here instead and remove the other checks.

File: test/vaadin-list-mixin.html [hunk 6]
```diff
@@ -168,10 +170,11 @@
     }
 
     describe('vaadin-list-mixin', () => {
-      let list;
+      let list, list2;
```
Comment (tomivirkki) [inline test/vaadin-list-mixin.html]: `list2` isn't very descriptive for a name. Rather `hiddenList`. If it makes sense, you could create a separate `describe('hidden-items', ...)` block for the tests with a list with hidden items.

File: vaadin-list-mixin.html (hunk 4 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done

File: vaadin-list-mixin.html (hunk 5 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done. I removed the checking condition on _focus(idx).

File: test/vaadin-list-mixin.html [hunk 7]
```diff
@@ -635,13 +637,31 @@
...
           expect(getComputedStyle(list.items[list.items.length - 1]).getPropertyValue('display')).to.equal('none');
           expect(list.items[list.items.length - 2].focused).to.be.true;
         });
+
+        it('should not set tabIndex=0 to hidden items, but the next one in the loop', () => {
+          list._focus(1);
+          [-1, -1, 0].forEach((val, idx) => expect(list.items[idx].tabIndex).to.be.equal(val));
+        });
+
+        it('key search should skip hidden items', () => {
```
Comment (web-padawan) [inline test/vaadin-list-mixin.html]: nit: I would rename the test to start with the word "should"

```js
it('should not include items to key search', () => {
```

File: vaadin-list-mixin.html [hunk 8]
```diff
@@ -208,8 +208,12 @@
       return -1;
     }
 
+    _isItemHidden(item) {
+      return getComputedStyle(item).getPropertyValue('display') === 'none';
```
Comment (web-padawan) [inline vaadin-list-mixin.html]: You can shorten this to `getComputedStyle(item).display`

File: test/vaadin-list-mixin.html (hunk 6 above)
Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: done

File: test/vaadin-list-mixin.html (hunk 7 above)
Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: done

File: vaadin-list-mixin.html (hunk 8 above)
Comment (hongduc-phan) [reply vaadin-list-mixin.html]: done

File: test/vaadin-list-mixin.html [hunk 9]
```diff
@@ -635,13 +637,31 @@
...
         it('should move focus to next not hidden element on "arrow-up"', () => {
           expect(list.items[0].focused).to.be.true;
           expect(getComputedStyle(list.items[0]).getPropertyValue('display')).to.equal('block');
           arrowUp(list);
           expect(getComputedStyle(list.items[list.items.length - 1]).getPropertyValue('display')).to.equal('none');
           expect(list.items[list.items.length - 2].focused).to.be.true;
         });
+
+        it('should not set tabIndex=0 to hidden items, but the next one in the loop', () => {
+          list._focus(1);
```
Comment (tomivirkki) [inline test/vaadin-list-mixin.html]: nit: Avoid using private APIs in the tests.

File: test/vaadin-list-mixin.html [hunk 10]
```diff
@@ -635,13 +637,31 @@
...
+          [-1, -1, 0].forEach((val, idx) => expect(list.items[idx].tabIndex).to.be.equal(val));
+        });
+
+        it('should skip hidden items to key search', () => {
+          keyDownChar(list, 'b');
+          keyDownChar(list, 'b');
+          expect(list.items[6].focused).to.be.true;
```
Comment (tomivirkki) [inline test/vaadin-list-mixin.html]: Rather than using magic numbers like 6 or 4 in the tests, prefer something that is more tolerant to changes in the fixture. Like `expect(list.items.filter(item => item.focused)[0].hidden).to.be.false;`

Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: fixed

File: test/vaadin-list-mixin.html (hunk 9 above)
Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: fixed!

File: test/vaadin-list-mixin.html (hunk 10 above)
Comment (tomivirkki) [reply test/vaadin-list-mixin.html]: You only fixed the other test. The test below still uses a magic number.

Comment (tomivirkki) [reply test/vaadin-list-mixin.html]: Discussed in person. Better to still test that the expected item is focused: `expect(list.items.filter(item => item.textContent === 'Pub')[0].focused).to.be.true;`

File: test/vaadin-list-mixin.html (hunk 9 above)
Comment (tomivirkki) [reply test/vaadin-list-mixin.html]: Now the tests pass even without this change https://github.com/vaadin/vaadin-list-mixin/pull/64/files#diff-9e82cf3b8875402729200020ea93243bL212-R216

See if something is not working correctly without the change. If everything works, the change isn't needed; if something is broken, fix the test to check the right things.

Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: hi !
I tried to use external API in tests:
function arrowDown(target) {
      MockInteractions.keyDownOn(target, 40, [], 'ArrowDown');
    }
. In source code, I checked hidden item in _setFocusable(idx) -> _focus(idx) -> _onKeydown(event).
The test will be failed if without checking hidden items. and not work correctly and we type first letter, the hidden items can be focused and tab index will be = 0.

File: test/vaadin-list-mixin.html (hunk 10 above)
Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: fixed!

File: test/vaadin-list-mixin.html (hunk 9 above)
Comment (tomivirkki) [reply test/vaadin-list-mixin.html]: I don't understand. The situation is still the same. The tests pass even without the change.

Comment (hongduc-phan) [reply test/vaadin-list-mixin.html]: I checked it and as you told me, I check the element is hidden in _setFocusable(idx) is not affected any. So I removed it.
